engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -191,6 +205,67 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     }
   }
 
+  @ExportMessage
+  int identityHashCode(
+      @Cached ConditionProfile isHashCodeComputedProfile,
+      @Cached("createIdentityProfile()") ValueProfile hashValueProfile
+  ) {
+    if (isHashCodeComputedProfile.profile(hashCode != null)) {
+      return hashValueProfile.profile(hashCode);
+    } else {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      hashCode = computeIdentityHashCode(this);
+      return hashCode;
+    }
+  }
+
+  @ExportMessage
+  static class IsIdenticalOrUndefined {
+    @Specialization
+    static TriState isIdenticalOrUndefined(
+        Atom thisAtom,
+        Atom otherAtom,
+        @Cached("createCountingProfile()") ConditionProfile thisAtomHasNotHash,
+        @Cached("createCountingProfile()") ConditionProfile otherAtomHasNotHash
+    ) {
+      if (thisAtomHasNotHash.profile(thisAtom.hashCode == null)) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        thisAtom.hashCode = computeIdentityHashCode(thisAtom);
+      }
+      int thisAtomHash = thisAtom.hashCode;
+      if (otherAtomHasNotHash.profile(otherAtom.hashCode == null)) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        otherAtom.hashCode = computeIdentityHashCode(otherAtom);
+      }
+      int otherAtomHash = otherAtom.hashCode;
+      return thisAtomHash == otherAtomHash ? TriState.TRUE : TriState.FALSE;
+    }
+
+    @Fallback
+    static TriState fallBack(Atom thisAtom, Object other) {
+      return TriState.FALSE;
+    }
+  }
+
+  @TruffleBoundary
+  static int computeIdentityHashCode(Atom atom) {
+    InteropLibrary interop = InteropLibrary.getUncached();
+    int[] hashCodes =  Arrays.stream(atom.fields)
```

Comment:
Please just `for` loop and `sum` field updated on each cycle - it is going to be more responsible towards GC - I am afraid with this stream magic, the compiler will be unable to eliminate allocation of `int[]`. But feel free to prove me being wrong.